src/render.rs

```diff
@@ -0,0 +1,123 @@
+use piet_gpu::Renderer;
+use piet_gpu_hal::{
+    CmdBuf, Error, ImageLayout, Instance, QueryPool, Semaphore, Session, SubmittedCmdBuf, Surface,
+    Swapchain,
+};
+use piet_scene::Scene;
+
+pub const NUM_FRAMES: usize = 2;
+
+pub struct PgpuState {
+    #[allow(unused)]
+    instance: Instance,
+    #[allow(unused)]
+    surface: Option<Surface>,
+    swapchain: Swapchain,
+    session: Session,
+    present_semaphores: Vec<Semaphore>,
+    query_pools: Vec<QueryPool>,
+    cmd_bufs: [Option<CmdBuf>; NUM_FRAMES],
+    submitted: [Option<SubmittedCmdBuf>; NUM_FRAMES],
+    renderer: Renderer,
+    current_frame: usize,
+}
+
+impl PgpuState {
+    pub fn new(
+        window: &dyn raw_window_handle::HasRawWindowHandle,
+        display: &dyn raw_window_handle::HasRawDisplayHandle,
+        width: usize,
+        height: usize,
+    ) -> Result<Self, Error> {
+        println!("size: {}, {}", width, height);
+        let instance = Instance::new(Default::default())?;
+        let surface = unsafe {
+            instance
+                .surface(display.raw_display_handle(), window.raw_window_handle())
+                .ok()
+        };
+        unsafe {
+            let device = instance.device()?;
+            let swapchain =
+                instance.swapchain(width, height, &device, surface.as_ref().unwrap())?;
+            let session = Session::new(device);
+            let present_semaphores = (0..NUM_FRAMES)
+                .map(|_| session.create_semaphore())
+                .collect::<Result<Vec<_>, Error>>()?;
+            let query_pools = (0..NUM_FRAMES)
+                .map(|_| session.create_query_pool(Renderer::QUERY_POOL_SIZE))
+                .collect::<Result<Vec<_>, Error>>()?;
+            let cmd_bufs: [Option<CmdBuf>; NUM_FRAMES] = Default::default();
+            let submitted: [Option<SubmittedCmdBuf>; NUM_FRAMES] = Default::default();
+            let renderer = Renderer::new(&session, width, height, NUM_FRAMES)?;
+            let current_frame = 0;
+            Ok(Self {
+                instance,
+                surface,
+                swapchain,
+                session,
+                present_semaphores,
+                query_pools,
+                cmd_bufs,
+                submitted,
+                renderer,
+                current_frame,
+            })
+        }
+    }
+
+    pub fn frame_index(&self) -> usize {
+        self.current_frame % NUM_FRAMES
+    }
+
+    pub fn pre_render(&mut self) -> Option<Vec<f64>> {
+        let frame_idx = self.frame_index();
+        if let Some(submitted) = self.submitted[frame_idx].take() {
+            self.cmd_bufs[frame_idx] = submitted.wait().unwrap();
+            Some(unsafe {
+                self.session
+                    .fetch_query_pool(&self.query_pools[frame_idx])
+                    .unwrap()
+            })
```

Comment:
I tried adding some prints around, and it seems to hang here on X11

Reply:
Did you update to latest piet-gpu, it sounds like https://github.com/linebender/piet-gpu/pull/207

Reply:
To do so, run `cargo update` before you run it. That solved the hang for me.